### PR TITLE
PS parameters

### PR DESCRIPTION
**Pull request by andreea-alexandru**
Changed logic of how to select the PS parameters given the degree of the polynomial that needs to be evaluated. In particular, for commonly used degrees (< 2205), hardcode parameters that minimize the multiplicative depth, and at that multiplicative depth, minimize the number of homomorphic multiplications. For larger degrees, a heuristic tries to approximate this logic.

### CODE REVIEW

File: src/pke/lib/scheme/ckksrns/ckksrns-utils.cpp [hunk 1]
```diff
@@ -233,31 +234,64 @@ std::shared_ptr<longDiv> LongDivisionChebyshev(const std::vector<double>& f, con
...
+    std::fill(mlist.begin() + 1007, mlist.begin() + 1083, 5); // n in [1008,1083], m = 5
+    std::fill(mlist.begin() + 1083, mlist.begin() + 2015, 6); // n in [1084,2015], m = 6
+    std::fill(mlist.begin() + 2015, mlist.begin() + 2031, 7); // n in [2016,2031], m = 7
+    std::fill(mlist.begin() + 2031, mlist.end(), 6); // n in [2032,2204], m = 6
+
+	if (  n <= upper_bound ) { // hard-coded values
+		uint32_t m = mlist[n-1];
+		uint32_t k = std::floor(n / ((1 << m) - 1)) + 1;
+
+		return std::vector<uint32_t>{{k,m}};
```
Comment (dsuponitskiy) [inline src/pke/lib/scheme/ckksrns/ckksrns-utils.cpp:274]: you don't need double brackets to create a vector in this case

File: src/pke/lib/scheme/ckksrns/ckksrns-utils.cpp [hunk 2]
```diff
@@ -233,31 +234,64 @@ std::shared_ptr<longDiv> LongDivisionChebyshev(const std::vector<double>& f, con
...
+					if(std::abs(std::floor(log2(k)) - std::floor(log2(sqrt(n/2)))) <= 1) {
+						klist.push_back(k); mlist2.push_back(m);
+						multlist.push_back(k + 2*m + (1 << (m-1)) - 4);
+					}
+				}
+			}
+		}
+		uint32_t minIndex = std::min_element(multlist.begin(),multlist.end()) - multlist.begin();
+
+		return std::vector<uint32_t>{{klist[minIndex],mlist2[minIndex]}};
```
Comment (dsuponitskiy) [inline src/pke/lib/scheme/ckksrns/ckksrns-utils.cpp:293]: you don't need double brackets to create a vector in this case

File: src/pke/lib/scheme/ckksrns/ckksrns-utils.cpp [hunk 3]
```diff
@@ -233,31 +234,64 @@ std::shared_ptr<longDiv> LongDivisionChebyshev(const std::vector<double>& f, con
...
-                    mlist.push_back(m);
-                }
-            }
-        }
-    }
-
-    uint32_t minIndex = std::min_element(mlist.begin(), mlist.end()) - mlist.begin();
-
-    return std::vector<uint32_t>{{klist[minIndex], mlist[minIndex]}};
+    const uint32_t upper_bound = 2204;
```
Comment (dsuponitskiy) [inline src/pke/lib/scheme/ckksrns/ckksrns-utils.cpp:249]: make upper_bound all capital

File: src/pke/lib/scheme/ckksrns/ckksrns-utils.cpp [hunk 4]
```diff
@@ -233,31 +234,64 @@ std::shared_ptr<longDiv> LongDivisionChebyshev(const std::vector<double>& f, con
...
+	else { // heuristic for larger degrees
+		std::vector<uint32_t> klist;
+		std::vector<uint32_t> mlist2;
+		std::vector<uint32_t> multlist;
+
+		for(uint32_t k = 1; k <= n; k++){
+			for(uint32_t m = 1; m <= std::ceil(log2(n / k) + 1) + 1; m ++){
+				if(int32_t(n) - int32_t(k * ((1 << m) - 1)) < 0){ 
+					if(std::abs(std::floor(log2(k)) - std::floor(log2(sqrt(n/2)))) <= 1) {
+						klist.push_back(k); mlist2.push_back(m);
```
[the remark this answers is not in the document]
Comment (andreea-alexandru) [reply src/pke/lib/scheme/ckksrns/ckksrns-utils.cpp]: done

File: src/pke/lib/scheme/ckksrns/ckksrns-utils.cpp (hunk 1 above)
Comment (andreea-alexandru) [reply src/pke/lib/scheme/ckksrns/ckksrns-utils.cpp]: done

File: src/pke/lib/scheme/ckksrns/ckksrns-utils.cpp (hunk 2 above)
Comment (andreea-alexandru) [reply src/pke/lib/scheme/ckksrns/ckksrns-utils.cpp]: done

File: src/pke/lib/scheme/ckksrns/ckksrns-utils.cpp (hunk 3 above)
Comment (andreea-alexandru) [reply src/pke/lib/scheme/ckksrns/ckksrns-utils.cpp]: done

**Review by dsuponitskiy — APPROVED** [review]
looks good

**Review by yspolyakov — APPROVED** [review]

**Comment by yspolyakov** [comment]
@andreea-alexandru Please also update the table in https://github.com/openfheorg/openfhe-development/blob/main/src/pke/examples/FUNCTION_EVALUATION.md

**Pull request merged**